packages/realm/src/Object.ts

```diff
@@ -382,17 +382,17 @@ export class RealmObject<T = DefaultObject, RequiredProperties extends keyof Omi
    * and [flatted](https://www.npmjs.com/package/flatted) to stringify Realm entities that have circular structures.
    * @returns A plain object.
    */
-  toJSON(_?: string, cache?: unknown): DefaultObject;
+  toJSON(_?: string, cache?: unknown): T | Partial<T>;
```

Comment:
Why the `Partial<T>`?

Also, a property of type `Realm.List` is expected to return an array of values, in which case, using the `T` directly would be incorrect.

Reply:
You're right, we need to mapping the array case. I'll check how can I improve it.

About the partial, when we are creating and mapping the values to create the object we have a partial of T, and we save in the cache, if some we have some break in the flow the cache will still be there (I assume) and the fn will return the partial of T, this is the mutabillity problem. Should I consider that we will never break and always return a complete object?

Reply:
> we will never break and always return a complete object?

I not 100% sure I follow. But what I think you're referring to is the fact that the object graph might contain cycles.
As I understand the `toJSON` semantics, it needs to transform objects into JSON serializable values, which might contain cycles and that's exactly what the implementation tries to do. I.e. if two objects refer to each other the cache would be used to detect the cycle and the resulting object would be two plain objects cyclicly referring to each other.

Reply:
I think should return `OmittedRealmTypes<T>`

In document https://www.mongodb.com/docs/atlas/device-sdks/sdk/react-native/model-data/define-a-realm-object-model/

```javascript
class Book extends Realm.Object<Book> {
  name!: string;
  price?: number;

  static schema: ObjectSchema = {
    name: 'Book',
    properties: {
      name: {type: 'string', indexed: true},
      price: 'int?',
    },
  };
}
```

`T` is `Book` (extends `Realm.Object`) so `T` has `Realm.Object`'s methods, properties

but `toJSON()` remove those methods, properties like `OmittedRealmTypes<T>`

I patched it locally to `toJSON(_?: string, cache?: unknown): OmittedRealmTypes<T>;` and didn't encounter any issues while using it.